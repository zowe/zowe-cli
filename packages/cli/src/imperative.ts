/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";

import { IImperativeConfig } from "@zowe/imperative";
import { Constants } from "./Constants";

import { ZosmfProfile } from "@zowe/zosmf-for-zowe-sdk";
import { ZosTsoProfile } from "@zowe/zos-tso-for-zowe-sdk";
import { ZosUssProfile } from "@zowe/zos-uss-for-zowe-sdk";
import { ProfileConstants } from "@zowe/core-for-zowe-sdk";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: [
        "daemon/*.definition!(.d).*s",
        "provisioning/*.definition!(.d).*s",
        "workflows/*.definition!(.d).*s",
        "zosconsole/*.definition!(.d).*s",
        "zosfiles/*.definition!(.d).*s",
        "zosjobs/*.definition!(.d).*s",
        "zosmf/*.definition!(.d).*s",
        "zostso/*.definition!(.d).*s",
        "zosuss/*.definition!(.d).*s",
        "zoslogs/*.definition!(.d).*s"
    ],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    webHelpLogoImgPath: path.join(__dirname, "..", "web-help-logo.png"),
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        }
    },
    apimlConnLookup: [
        {
            apiId: "ibm.zosmf",
            gatewayUrl: "api/v1",
            connProfType: "zosmf"
        }
    ],
    baseProfile: {
        ...ProfileConstants.BaseProfile,
        authConfig: [
            {
                serviceName: "apiml",
                handler: __dirname + "/auth/ApimlAuthHandler",
                login: {
                    summary: ProfileConstants.APIML_LOGIN_SUMMARY,
                    description: ProfileConstants.APIML_LOGIN_DESCRIPTION,
                    examples: [
                        ProfileConstants.APIML_LOGIN_EXAMPLE1,
                        ProfileConstants.APIML_LOGIN_EXAMPLE2
                    ],
                    options: [
                        ProfileConstants.BASE_OPTION_HOST,
                        ProfileConstants.BASE_OPTION_PORT,
                        ProfileConstants.BASE_OPTION_USER,
                        ProfileConstants.BASE_OPTION_PASSWORD,
                        ProfileConstants.BASE_OPTION_REJECT_UNAUTHORIZED,
                        ProfileConstants.BASE_OPTION_CERT_FILE,
                        ProfileConstants.BASE_OPTION_CERT_KEY_FILE,
                        ProfileConstants.BASE_OPTION_CERT_ACCOUNT,
                        ProfileConstants.BASE_OPTION_CERT_KEY_ACCOUNT
                    ]
                },
                logout: {
                    summary: ProfileConstants.APIML_LOGOUT_SUMMARY,
                    description: ProfileConstants.APIML_LOGOUT_DESCRIPTION,
                    examples: [
                        ProfileConstants.APIML_LOGOUT_EXAMPLE1,
                        ProfileConstants.APIML_LOGOUT_EXAMPLE2
                    ],
                    options: [
                        ProfileConstants.BASE_OPTION_HOST,
                        ProfileConstants.BASE_OPTION_PORT,
                        ProfileConstants.APIML_LOGOUT_OPTION_TOKEN_TYPE,
                        ProfileConstants.BASE_OPTION_TOKEN_VALUE,
                        ProfileConstants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                }
            }
        ]
    },
    authGroupConfig: {
        authGroup: {
            summary: ProfileConstants.AUTH_GROUP_SUMMARY,
            description: ProfileConstants.AUTH_GROUP_DESCRIPTION
        }
    },
    configAutoInitCommandConfig: {
        handler: __dirname + "/config/auto-init/ApimlAutoInitHandler",
        provider: "APIML",
        autoInit: {
            options: [
                ProfileConstants.AUTO_INIT_OPTION_HOST,
                ProfileConstants.AUTO_INIT_OPTION_PORT,
                ProfileConstants.AUTO_INIT_OPTION_USER,
                ProfileConstants.AUTO_INIT_OPTION_PASSWORD,
                ProfileConstants.AUTO_INIT_OPTION_REJECT_UNAUTHORIZED,
                ProfileConstants.AUTO_INIT_OPTION_TOKEN_TYPE,
                ProfileConstants.AUTO_INIT_OPTION_TOKEN_VALUE,
                ProfileConstants.AUTO_INIT_OPTION_CERT_FILE,
                ProfileConstants.AUTO_INIT_OPTION_CERT_KEY_FILE,
                ProfileConstants.BASE_OPTION_CERT_ACCOUNT,
                ProfileConstants.BASE_OPTION_CERT_KEY_ACCOUNT
            ]
        },
        profileType: "base"
    },
    profiles: [
        ZosmfProfile,
        ZosTsoProfile,
        ZosUssProfile
    ]
};
module.exports = config;
