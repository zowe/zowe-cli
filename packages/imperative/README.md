# Imperative CLI Framework

[![codecov](https://codecov.io/gh/zowe/imperative/branch/master/graph/badge.svg)](https://codecov.io/gh/zowe/imperative)
[![OpenSSF Best Practices](https://bestpractices.coreinfrastructure.org/projects/2245/badge)](https://bestpractices.coreinfrastructure.org/projects/2245)

Imperative CLI Framework is a command processing system that lets you quickly build customized command-line interfaces. Prior to V3 of the Zowe CLI, Imperative was a seperate package. It is now fully integrated with the CLI where it continues to provide groundwork for CLI infrastrure, functionality and plug-ins.
## Software Requirements

- [**Install Node.js package manager**](https://nodejs.org/en/download/package-manager) on your computer. Node.jsÂ® is a JavaScript runtime environment on which we architected Imperative CLI Framework.

- You must have a means to execute ".sh" (bash) scripts to run integration tests. On Windows, you can install "Git Bash", which is bundled with the standard [Git](https://git-scm.com/downloads) installation - (choose the "Use Git and Unix Tools from Windows Command Prompt" installation option). When you run the integration tests on Windows, you must have Administrative authority to enable the integration tests to create symbolic links.

> [!WARNING]
> Broadcom Inc. does not maintain the prerequisite software that Imperative CLI Framework requires. You are responsible for updating Node.js and other prerequisites on your computer. We recommend that you update Node.js regularly to the latest Long Term Support (LTS) version.

## Install Imperative as a Dependency

Issue the following commands to install Imperative CLI Framework as a dependency.

- **Install `@latest` version:**

    Be aware that if you update via `@latest`, you accept breaking changes into your project.

    ``` bash
    npm install @zowe/imperative
    ```

- **Install `@zowe-v3-lts` version:**

    This is a Long Term Support release that is guaranteed to have no breaking changes.

    ``` bash
    npm install @zowe/imperative@zowe-v3-lts
    ```

> [!NOTE]
> If you want to install the bleeding edge version of Imperative, you can append
> `--@zowe:registry=https://zowe.jfrog.io/zowe/api/npm/npm-release/` to the install command to get it from a staging registry. It is not recommended to use this registry for production dependencies.

## Build, Test and Expand Imperative

### Build and Install Imperative CLI Framework from Source
To build and install the Imperative CLI Framework, follow these steps:

1. Clone the `zowe/imperative` project to your PC.
2. From the command line, issue `cd [relative path]/imperative`
3. Issue `npm install`
4. Issue `npm run build`
5. Issue `npm run test`

To build the entire project (including test stand-alone CLIs):
`npm run build`

To build only imperative source:
`npm run build:packages`

### Run Tests
Command | Description
--- | ---
`npm run test` | Run all automated tests (unit & integration)
`npm test:unit` | Run unit tests
`npm test:integration` | Run integration tests
`npm test:system` | Run system tests (requires IPv6 connection)

> [!NOTE]
> To build and install the test CLIs used by the integration tests:
> 1. `node scripts/sampleCliTool.js build`
> 2. `node scripts/sampleCliTool.js install`

### Sample Applications

We provide a sample plug-in that you can use to get started developing your own plug-ins. See the [Zowe CLI Sample Plug-in](https://github.com/zowe/zowe-cli-sample-plugin).

## Documentation
For historic documentation that describes how to define commands, work with user profiles, and more, see [Imperative CLI Framework wiki](https://github.com/zowe/imperative/wiki).

## Contribute
For information about how you can contribute code to Imperative CLI Framework, see [CONTRIBUTING](../../CONTRIBUTING.md).

## Submit an Issue
To report a bug or request an enhancement, please [submit an issue](https://github.com/zowe/zowe-cli/issues/new/choose).

## Versioning
Imperative CLI Framework uses Semantic Versioning (SemVer) for versioning. For more information, see the [Semantic Versioning](https://semver.org/) website.

## Licensing Imperative CLI Framework
For Imperative CLI Framework licensing rules, requirements, and guidelines, see [LICENSE](../../LICENSE).
