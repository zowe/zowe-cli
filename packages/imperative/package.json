{
  "name": "@zowe/imperative",
  "version": "8.0.0-next.202402271901",
  "description": "framework for building configurable CLIs",
  "author": "Zowe",
  "license": "EPL-2.0",
  "homepage": "https://github.com/zowe/zowe-cli/tree/master/packages/imperative#readme",
  "bugs": {
    "url": "https://github.com/zowe/zowe-cli/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/zowe/zowe-cli.git"
  },
  "keywords": [
    "CLI",
    "framework",
    "zowe"
  ],
  "files": [
    "lib",
    "web-help/dist",
    "web-diff"
  ],
  "publishConfig": {
    "registry": "https://zowe.jfrog.io/zowe/api/npm/npm-local-release/"
  },
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "scripts": {
    "build:packages": "tsc --pretty",
    "build": "npm run build:packages && npm run build:webHelp",
    "lint": "eslint \"packages/**/*.ts\" \"**/__tests__/**/*.ts\"",
    "lint:packages": "eslint \"packages/**/*.ts\" --ignore-pattern \"**/__tests__/**/*.ts\"",
    "lint:tests": "eslint \"**/__tests__/**/*.ts\"",
    "circularDependencyCheck": "madge -c lib",
    "watch": "concurrently \"tsc --pretty --watch\" \"npm run watch:webHelp\"",
    "build:webHelp": "cd web-help && npm run build",
    "watch:webHelp": "cd web-help && npm run watch",
    "typedoc": "typedoc --options ./typedoc.json ./src/",
    "typedocSpecifySrc": "typedoc --options ./typedoc.json",
    "prepack": "node ../../scripts/prepareLicenses.js",
    "clean": "rimraf lib tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "@types/yargs": "^17.0.32",
    "chalk": "^4.1.2",
    "cli-table3": "^0.6.3",
    "comment-json": "~4.2.3",
    "cross-spawn": "^7.0.3",
    "dataobject-parser": "^1.2.25",
    "deepmerge": "^4.3.1",
    "diff": "^5.2.0",
    "diff2html": "3.4.20-usewontache.1.60e7a2e",
    "fast-glob": "^3.3.2",
    "fastest-levenshtein": "^1.0.16",
    "find-up": "^5.0.0",
    "fs-extra": "^11.0.0",
    "jest-diff": "^29.0.0",
    "js-yaml": "^4.1.0",
    "jsonfile": "^6.0.0",
    "jsonschema": "^1.4.1",
    "lodash": "^4.17.21",
    "lodash-deep": "^2.0.0",
    "log4js": "^6.9.1",
    "markdown-it": "^12.3.2",
    "mustache": "^4.2.0",
    "npm-package-arg": "^11.0.1",
    "opener": "^1.5.2",
    "pacote": "^17.0.6",
    "prettyjson": "^1.2.5",
    "progress": "^2.0.3",
    "read": "^3.0.1",
    "semver": "^7.5.2",
    "stack-trace": "^0.0.10",
    "strip-ansi": "^6.0.1",
    "which": "^4.0.0",
    "wrap-ansi": "^7.0.0",
    "yamljs": "^0.3.0",
    "yargs": "^17.7.2"
  },
  "devDependencies": {
    "@types/cross-spawn": "^6.0.6",
    "@types/diff": "^5.0.9",
    "@types/fs-extra": "^11.0.4",
    "@types/glob": "^8.1.0",
    "@types/jsonfile": "^6.1.4",
    "@types/lodash": "^4.14.202",
    "@types/lodash-deep": "^2.0.4",
    "@types/mustache": "^4.2.5",
    "@types/npm-package-arg": "^6.1.4",
    "@types/pacote": "^11.1.8",
    "@types/progress": "^2.0.7",
    "@types/stack-trace": "^0.0.29",
    "@zowe/secrets-for-zowe-sdk": "8.0.0-next.202402271901",
    "concurrently": "^8.0.0",
    "cowsay": "^1.6.0",
    "deep-diff": "^1.0.0",
    "get-function-arguments": "^1.0.0",
    "serve": "^14.2.1",
    "stream-to-string": "^1.2.0",
    "uuid": "^9.0.1",
    "web-help": "file:web-help",
    "yargs-parser": "^21.1.1"
  },
  "engines": {
    "node": ">=16.7.0"
  }
}
