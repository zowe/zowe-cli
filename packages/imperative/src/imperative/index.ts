/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./env/EnvironmentalVariableSettings";
export * from "./doc/IApimlSvcAttrs";
export * from "./doc/IDaemonContext";
export * from "./doc/IImperativeEnvironmentalVariableSetting";
export * from "./doc/IImperativeEnvironmentalVariableSettings";
export * from "./doc/IImperativeConfig";
export * from "./doc/IImperativeLoggingConfig";
export * from "./ConfigurationLoader";
export * from "./ConfigurationValidator";
export * from "./DefinitionTreeResolver";
export * from "./Imperative";
export * from "./LoggingConfigurer";
export * from "./plugins/AbstractPluginLifeCycle";
export * from "./plugins/PluginManagementFacility";
export * from "./auth/doc/IAuthHandlerApi";
export * from "./auth/handlers/AbstractAuthHandler";
export * from "./auth/handlers/BaseAuthHandler";
export * from "./config/cmd/auto-init/handlers/BaseAutoInitHandler";
