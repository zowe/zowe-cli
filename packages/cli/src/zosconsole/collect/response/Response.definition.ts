/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandDefinition } from "@zowe/imperative";
import { Constants } from "../../../Constants";

export const SyncResponseCommandDefinition: ICommandDefinition = {
    name: "sync-responses",
    aliases: ["sr"],
    summary: "Collect outstanding synchronous console response messages",
    description: `The z/OSMF console REST APIs return a "solicited response key" after successfully issuing ` +
    `a synchronous console command that produces solicited responses. You can use the "solicited response key"` +
    `on the "sync-responses" command to collect any additional outstanding solicited responses ` +
    `from the console the command was issued.` +
    `\n\n` +
    `In general, when issuing a z/OS console command, z/OS applications route responses to the originating console. ` +
    `The command response messages are referred to as "solicited command responses" (i.e. direct responses to the command issued). ` +
    `When issuing a z/OS console command using ${Constants.DISPLAY_NAME}, collection of all solicited command responses ` +
    `is attempted by default. However, there is no z/OS mechanism that indicates the total number of response messages that may be produced ` +
    `from a given command. Therefore, the ${Constants.DISPLAY_NAME} console APIs return a "solicited response key" ` +
    `that can be used to "follow-up" and collect any additional solicited command responses.`,
    type: "command",
    handler: __dirname + "/Response.handler",
    profile: {
        optional: ["zosmf"]
    },
    positionals: [
        {
            name: "responsekey",
            type: "string",
            description: `The "solicited response key" provided in response to a previously issued console command. ` +
            `Used by the z/OSMF console API to collect any additional outstanding solicited responses from a ` +
            `previously issued console command.`,
            regex: "^[a-zA-Z0-9]+$",
            required: true
        }
    ],
    options: [
        {
            name: "console-name",
            aliases: ["cn", "c"],
            description: `The name of the z/OS extended MCS console to direct the command. ` +
            `You must have the required authority to access the console specified. ` +
            `You may also specify an arbitrary name, if your installation allows dynamic creation ` +
            `of consoles with arbitrary names.`,
            type: "string",
            allowableValues: {
                values: [/^[a-zA-Z0-9]+$/.source],
                caseSensitive: false
            }
        }
    ],
    examples: [
        {
            description: "Collect any outstanding additional solicited response messages",
            options: "C4866969"
        }
    ]
};
