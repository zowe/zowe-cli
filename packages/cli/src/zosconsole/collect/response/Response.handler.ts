/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IHandlerParameters } from "@zowe/imperative";
import { CollectCommand, ICollectParms, IConsoleResponse } from "@zowe/zos-console-for-zowe-sdk";
import { ZosmfBaseHandler } from "@zowe/zosmf-for-zowe-sdk";

/**
 * Handle to collect a MVS console command response
 * @export
 * @class Handler
 * @implements {ICommandHandler}
 */
export default class Handler extends ZosmfBaseHandler {

    public async processCmd(commandParameters: IHandlerParameters) {

        const collectParms: ICollectParms = {
            commandResponseKey: commandParameters.arguments.responsekey,
            consoleName: commandParameters.arguments["console-name"]
        };

        const response: IConsoleResponse = await CollectCommand.collect(this.mSession, collectParms);

        // Print out the response
        commandParameters.response.console.log(response.commandResponse);

        // Return as an object when using --response-format-json
        commandParameters.response.data.setObj(response);

    }
}
