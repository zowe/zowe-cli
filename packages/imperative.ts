/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

// WIP Imperative version of Brightside
import { IImperativeConfig } from "@brightside/imperative";
import { Constants } from "./Constants";
import { ZosmfSession } from "./zosmf";
import {
    TSO_OPTION_ACCOUNT,
    TSO_OPTION_CHAR_SET,
    TSO_OPTION_CODE_PAGE,
    TSO_OPTION_COLUMNS,
    TSO_OPTION_LOGON_PROCEDURE,
    TSO_OPTION_REGION_SIZE,
    TSO_OPTION_ROWS
} from "./zostso/src/cli/constants/ZosTso.constants";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: ["**/cli/*.definition!(.d).*s"],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        },
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_HOST,
                    },
                    port: {
                        type: "number",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PORT,
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_USER,
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PASSWORD,
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_REJECT_UNAUTHORIZED,
                    },
                    basePath: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_BASE_PATH,
                    },
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                },
                {
                    options: "zosAPIML --host zosAPIML --port 2020 --user ibmuser --password myp4ss --reject-unauthorized false --base-path basePath",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                }
            ],
            updateProfileExamples: [
                {
                    options: "zos123 --user newuser --password newp4ss",
                    description: "Update a zosmf profile named 'zos123' with a new username and password"
                }
            ]
        },
        {
            type: "tso",
            schema: {
                type: "object",
                title: "TSO Profile",
                description: "z/OS TSO/E User Profile",
                properties: {
                    account: {
                        type: "string",
                        optionDefinition: TSO_OPTION_ACCOUNT,
                    },
                    characterSet: {
                        type: "string",
                        optionDefinition: TSO_OPTION_CHAR_SET,
                    },
                    codePage: {
                        type: "string",
                        optionDefinition: TSO_OPTION_CODE_PAGE,
                    },
                    columns: {
                        type: "number",
                        optionDefinition: TSO_OPTION_COLUMNS,
                    },
                    logonProcedure: {
                        type: "string",
                        optionDefinition: TSO_OPTION_LOGON_PROCEDURE,
                    },
                    regionSize: {
                        type: "number",
                        optionDefinition: TSO_OPTION_REGION_SIZE,
                    },
                    rows: {
                        type: "number",
                        optionDefinition: TSO_OPTION_ROWS,
                    },
                },
                required: ["account"],
            },
            createProfileExamples: [
                {
                    description: "Create a tso profile called 'myprof' with default settings and JES accounting information of 'IZUACCT'",
                    options: "myprof -a IZUACCT"
                },
                {
                    description: "Create a tso profile called 'largeregion' with a region size of 8192, a logon procedure of MYPROC, and " +
                        "JES accounting information of '1234'",
                    options: "largeregion -a 1234 --rs 8192"
                }
            ],
            updateProfileExamples: [
                {
                    description: "Update a tso profile called myprof with new JES accounting information",
                    options: "myprof -a NEWACCT"
                }
            ]
        },
    ]
};
module.exports = config;
