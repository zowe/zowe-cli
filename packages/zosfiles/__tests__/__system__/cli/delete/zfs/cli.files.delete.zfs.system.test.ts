/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Session } from "@zowe/imperative";
import { getUniqueDatasetName, runCliScript, stripNewLines } from "../../../../../../../__tests__/__src__/TestUtils";
import { TestEnvironment } from "../../../../../../../__tests__/__src__/environment/TestEnvironment";
import { ITestEnvironment } from "../../../../../../../__tests__/__src__/environment/doc/response/ITestEnvironment";
import { ITestPropertiesSchema } from "../../../../../../../__tests__/__src__/properties/ITestPropertiesSchema";

const ZOWE_OPT_BASE_PATH = "ZOWE_OPT_BASE_PATH";

let REAL_SESSION: Session;
// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;
let TEST_ENVIRONMENT_NO_PROF: ITestEnvironment;
let defaultSystem: ITestPropertiesSchema;
let fsname: string;
let volume: string;

describe("Delete z/OS File System", () => {
    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            tempProfileTypes: ["zosmf"],
            testName: "zos_delete_zos_file_system"
        });

        defaultSystem = TEST_ENVIRONMENT.systemTestProperties;

        REAL_SESSION = TestEnvironment.createZosmfSession(TEST_ENVIRONMENT);

        fsname = getUniqueDatasetName(defaultSystem.zosmf.user);
        volume = defaultSystem.datasets.vol;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    describe("without profiles", () => {

        let defaultSys: ITestPropertiesSchema;

        // Create the unique test environment
        beforeAll(async () => {
            TEST_ENVIRONMENT_NO_PROF = await TestEnvironment.setUp({
                testName: "zos_files_delete_zos_file_system_without_profile"
            });

            defaultSys = TEST_ENVIRONMENT_NO_PROF.systemTestProperties;
        });

        afterAll(async () => {
            await TestEnvironment.cleanUp(TEST_ENVIRONMENT_NO_PROF);
        });

        it("should delete a data set", async () => {
            // if API Mediation layer is being used (basePath has a value) then
            // set an ENVIRONMENT variable to be used by zowe.
            if (defaultSys.zosmf.basePath != null) {
                TEST_ENVIRONMENT_NO_PROF.env[ZOWE_OPT_BASE_PATH] = defaultSys.zosmf.basePath;
            }

            let response = runCliScript(__dirname + "/__scripts__/command/command_create_zfs_fully_qualified.sh",
                TEST_ENVIRONMENT_NO_PROF, [fsname, volume,
                    defaultSys.zosmf.host,
                    defaultSys.zosmf.port,
                    defaultSys.zosmf.user,
                    defaultSys.zosmf.pass]);

            response = runCliScript(__dirname + "/__scripts__/command/command_delete_zfs_fully_qualified.sh",
                TEST_ENVIRONMENT_NO_PROF, [fsname, "--for-sure",
                    defaultSys.zosmf.host,
                    defaultSys.zosmf.port,
                    defaultSys.zosmf.user,
                    defaultSys.zosmf.pass]);

            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toMatchSnapshot();
        });
    });

    describe("Success scenarios", () => {
        it("should delete a VSAM data set", async () => {
            let response = runCliScript(__dirname + "/__scripts__/command/command_create_zfs.sh",
                TEST_ENVIRONMENT, [fsname, volume]);

            response = runCliScript(__dirname + "/__scripts__/command/command_delete_zfs.sh",
                TEST_ENVIRONMENT, [fsname, "--for-sure"]);

            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toMatchSnapshot();
        });
    });

    describe("Expected failures", () => {
        it("should fail deleting a data set that does not exist", async () => {
            const notExistZfs = `${fsname}.NOTEXIST`
            const response = runCliScript(__dirname + "/__scripts__/command/command_delete_zfs.sh",
                TEST_ENVIRONMENT, [notExistZfs, "--for-sure"]);
            expect(response.status).toBe(1);
            expect(stripNewLines(response.stderr.toString())).toContain(`ENTRY ${notExistZfs} NOT DELETED`);
        });
    });
});
