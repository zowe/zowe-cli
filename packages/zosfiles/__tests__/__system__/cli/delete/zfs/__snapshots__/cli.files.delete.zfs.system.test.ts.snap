// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete z/OS File System Success scenarios should delete a VSAM data set 1`] = `
"================Z/OS FILES DELETE Z/OS FILE SYSTEM===============
z/OS file system deleted successfully.
"
`;

exports[`Delete z/OS File System without profiles should delete a data set 1`] = `
"================Z/OS FILES DELETE Z/OS FILE SYSTEM===============
z/OS file system deleted successfully.
"
`;
